course-design.md

```diff
@@ -0,0 +1,214 @@
+    - Idea: Shiny app here?
+
+- Version control: Using Git and GitHub as a sole user (part 1)
+    - Using Git in RStudio (standard add-commit-history)
+    - Using Git in the Terminal of RStudio (moving in history with checkout,
+    creating branches, adding and updating remotes)
+    - Setting up GitHub for R package (make use of usethis), pushing and pulling
+        - E.g. with pr_* functions from usethis
+
+- Checking correctness of code
+    - Using basic debugging strategies ("breakpoints", simple `print()` or `stop()` calls, `browser()`)
+    - Unit test (with testthat)
+    - Code coverage (with covr)
+    - With devtools
+
+- Build management
+    - With devtools
+    - General workflow (load_all, test, check)
+    - Maybe with drake?
+        - Cover it minimally or as a later stand-alone topic.
+        - Drake is (maybe?) more targeted to data analysis pipelines, not package dev
+    - Pre-push running test, build, and check
+    - Continuous integration using GitHub Actions 
+
+- Developing documentation and tutorials on usage
+    - Vignettes, README (with rmarkdown and usethis)
+    - Function documentation (with roxygen2), part 2
+    - learnr tutorials via `use_tutorials()`?
+    - Running spell checks and styling (with spelling and styler)
+
+- Exposing your package to the world with a website:
+    - Setting up a website (with pkgdown)
+    - Customizing the themes
+    - Getting all resources and material organized
+    - Defining lifecycle of your package and individual functions (with lifecycle)
+
+- Involving the community:
+    - Designing your package to be used by or contributed to from the community
+    (refer to Mozilla Science Labs Open Leaders material)
+    - Contributing guidelines 
+    - Being inclusive (code of conduct)
+    - Licensing
+    - Governance
+    - NEWS file
+
+- Version control: Workflows around using GitHub (part 2)
+    - Make use of R builtin tools like the usethis pull request helpers 
+    (and learning how to make use of this yourself)
+    - Using this workflow as a team or as a contributor
+
+- Project management of your package development (either sole or in a team)
+    - Issue tracking
+    - Labeling
+    - Project boards
+    - Milestones
+
+- Preparing for package release
+    - Check builds on other systems (with rhub)
+    - Running local CRAN checks
+    - Managing your versions, git tagging
+
+- R package development in a team-based environment
+    - Making use of GitHub Organizations/Teams
+    - Branch protection, member roles
+    - Resolving conflict, personality differences
+    - Running meetings, code of conduct
+    - Decision making
+    - Governance
+    - Issue assignment
+    - Relying on and using code of conduct to build culture and standards
+
+## Final projects for formal assessment
+
+Two project deliverables for learners:
+
+- Package created alone as the first deliverable
+    - Due after "package release" lesson
```

Comment:
I really like the rationale here, that you need to build a package alone to know how all the pieces fit together, then develop in a group, but I'm worried about how much content there is before the "Package release" lesson.  

I think once we have an idea of what this simpler package looks like, we might either streamline the content before this assessment to focus on just what you need to build it (i.e. if the package just need to pass `check()`, we could skip CI) and return to topics later for more detail.  Or, we can just indicate up front to an instructor the "must cover" sections and let them plan out a path to this assessment.

Reply:
> "how much content there is before the "Package release" lesson."

Yea, I was thinking leading up to this point there be several "milestone" deliverables to get to this point, so learners aren't left until the last moment to have this package built. But yes, you're right, maybe move this due date to right before "Exposing your package" section, at the "Build Management" section like you suggest. I put a comment about moving the local CRAN check to that Build section.

> "'must cover' section"

I like this idea. Including something like this in the "Instructions to Instructors" could let us modify the content to be extensible to shorter or longer semesters/workshops. I've added a comment in #10 about it.

Reply:
I also like the idea of getting to a published project/package quicker and then improving it with CI etc. I think that can be more more motivating for the learners, and also give more insight into why some of the meta aspects are needed once they have seen how it looks without them briefly.